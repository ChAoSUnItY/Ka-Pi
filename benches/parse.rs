use criterion::{black_box, criterion_group, criterion_main, Criterion};
use ka_pi::parse::to_class;

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("parse main", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/Main.class"
    )))));

    c.bench_function("parse enum", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/Enum.class"
    )))));

    c.bench_function("parse record", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/Record.class"
    )))));

    c.bench_function("parse visible annotation", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/VisibleAnnotation.class"
    )))));

    c.bench_function("parse invisible annotation", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/InvisibleAnnotation.class"
    )))));

    c.bench_function("parse annotation target", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/AnnotationTarget.class"
    )))));
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
