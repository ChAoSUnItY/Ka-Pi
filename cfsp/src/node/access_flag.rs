use bitflags::bitflags;

use serde::{Deserialize, Serialize};

bitflags! {
    /// Access flag for [node::class::Class].
    ///
    /// See [Table 4.1-B](https://docs.oracle.com/javase/specs/jvms/se20/jvms20.pdf#page=85).
    #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
    pub struct ClassAccessFlag: u16 {
        const Public = 0x0001;
        const Final = 0x0010;
        const Super = 0x0020;
        const Interface = 0x0200;
        const Abstract = 0x0400;
        const Synthetic = 0x1000;
        const Annotation = 0x2000;
        const Enum = 0x4000;
        const Module = 0x8000;
    }

    /// Access flag for [node::field::Field].
    ///
    /// See [Table 4.5-A](https://docs.oracle.com/javase/specs/jvms/se20/jvms20.pdf#page=110).
    #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
    pub struct FieldAccessFlag: u16 {
        const Public = 0x0001;
        const Private = 0x0002;
        const Protected = 0x0004;
        const Static = 0x0008;
        const Final = 0x0010;
        const Volatile = 0x0040;
        const Transient = 0x0080;
        const Synthetic = 0x1000;
        const Enum = 0x4000;
    }

    /// Access flag for [node::method::Method].
    ///
    /// See [Table 4.6-A](https://docs.oracle.com/javase/specs/jvms/se20/jvms20.pdf#page=112).
    #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
    pub struct MethodAccessFlag: u16 {
        const Public = 0x0001;
        const Private = 0x0002;
        const Protected = 0x0004;
        const Static = 0x0008;
        const Final = 0x0010;
        const Synchronized = 0x0020;
        const Bridge = 0x0040;
        const Varargs = 0x0080;
        const Native = 0x0100;
        const Abstract = 0x0400;
        const Strict = 0x0800;
        const Synthetic = 0x1000;
    }

    /// Access flag for [node::attribute::InnerClass].
    ///
    /// See [Table 4.7.6-A](https://docs.oracle.com/javase/specs/jvms/se20/jvms20.pdf#page=138).
    #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
    pub struct NestedClassAccessFlag: u16 {
        const Public = 0x0001;
        const Private = 0x0002;
        const Protected = 0x0004;
        const Static = 0x0008;
        const Final = 0x0010;
        const Interface = 0x0200;
        const Abstract = 0x0400;
        const Synthetic = 0x1000;
        const Annotation = 0x2000;
        const Enum = 0x4000;
    }

    /// Access flag for [node::attribute::MethodParameter].
    ///
    /// See [here](https://docs.oracle.com/javase/specs/jvms/se20/jvms20.pdf#page=183).
    #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
    pub struct ParameterAccessFlag: u16 {
        const Final = 0x0010;
        const Synthetic = 0x1000;
        const Mandated = 0x8000;
    }

    /// Access flag for [node::attribute::Module].
    ///
    /// See [here](https://docs.oracle.com/javase/specs/jvms/se20/jvms20.pdf#page=186).
    #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
    pub struct ModuleAccessFlag: u16 {
        const Open = 0x0020;
        const Synthetic = 0x1000;
        const Mandated = 0x8000;
    }

    /// Access flag for [node::attribute::module::Requires].
    ///
    /// See [here](https://docs.oracle.com/javase/specs/jvms/se20/jvms20.pdf#page=187).
    #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
    pub struct RequiresAccessFlag: u16 {
        const Transitive = 0x0020;
        const StaticPhase = 0x0040;
        const Synthetic = 0x1000;
        const Mandated = 0x8000;
    }

    /// Access flag for [node::attribute::module::Exports].
    ///
    /// See [here](https://docs.oracle.com/javase/specs/jvms/se20/jvms20.pdf#page=188).
    #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
    pub struct ExportsAccessFlag: u16 {
        const Synthetic = 0x1000;
        const Mandated = 0x8000;
    }

    /// Access flag for [node::attribute::module::Opens].
    ///
    /// See [here](https://docs.oracle.com/javase/specs/jvms/se20/jvms20.pdf#page=189).
    #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
    pub struct OpensAccessFlag: u16 {
        const Synthetic = 0x1000;
        const Mandated = 0x8000;
    }
}
