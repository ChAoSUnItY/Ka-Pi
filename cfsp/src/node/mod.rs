//! `node` module contains all specification-described data structures from
//! [The JVM Specification - Java SE 20 Edition](https://docs.oracle.com/javase/specs/jvms/se20/jvms20.pdf).
//!
//! Some of the structures are made to be user-friendly, at the same time makes it much more
//! straightforward to use.

pub mod access_flag;
pub mod attribute;
pub mod class;
pub mod constant;
mod error;
pub mod field;
pub mod method;
pub mod opcode;
pub mod signature;
