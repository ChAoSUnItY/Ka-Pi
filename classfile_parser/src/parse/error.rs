use std::io;

use thiserror::Error;

use crate::node::opcode::Opcode;

pub type ParseResult<T> = Result<T, ParseError>;

#[derive(Debug, Error)]
pub enum ParseError {
    #[error(transparent)]
    Io(#[from] io::Error),
    #[error("expected [0xCA, 0xFE, 0xBA, 0xBE] at the header of class file but got {0:?}")]
    MismatchedMagicNumber([u8; 4]),
    #[error("attempt to match {0} in {1:?} but got {2}")]
    MatchOutOfBoundUsize(&'static str, Vec<&'static str>, usize),
    #[error("attempt to match {0} in {1:?} but got '{2}'")]
    MatchOutOfBoundChar(&'static str, Vec<char>, char),
    #[error("attempt to match opcode but got opcode value {0}")]
    MatchOutOfBoundOpcode(u8),
    #[error("attempt to match sub opcode in wide opcode but got {0}, while {1:?} is allowed as sub opcode of wide opcode")]
    MatchOutOfBoundWideOpcode(u8, Vec<Opcode>),
    #[error("the parse for segment is finished but still remains {0} bytes")]
    Remains(usize),
    #[error("expected {1:?} but got '{0}'")]
    MismatchedCharacter(char, Vec<char>),
    #[error("attempt to parse {0} but is incomplete")]
    OutOfBound(&'static str),
}
