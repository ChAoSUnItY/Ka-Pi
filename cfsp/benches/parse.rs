use cfsp::parse::{to_class, ParsingOption};
use std::fs;
use std::io::Cursor;

use criterion::{criterion_group, criterion_main, BatchSize, BenchmarkId, Criterion, Throughput};

fn parsing_classes(c: &mut Criterion) {
    let mut g = c.benchmark_group("parse class");

    let compiled_source_folder =
        fs::read_dir("../compiled_source/out/production/compiled_source").unwrap();

    for entry in compiled_source_folder {
        let entry = entry.as_ref().unwrap().path();
        let class_name = entry.file_name().unwrap().to_str().unwrap();
        let class_bytes = fs::read(&entry).unwrap();

        g.throughput(Throughput::Bytes(class_bytes.len() as u64));
        g.bench_with_input(
            BenchmarkId::from_parameter(class_name),
            &class_bytes,
            |b, bytes| {
                b.iter_batched(
                    || Cursor::new(bytes),
                    |mut cursor| {
                        to_class(&mut cursor, ParsingOption::default().parse_attribute())
                            .expect("Parsing fails on benchmarking");
                    },
                    BatchSize::SmallInput,
                );
            },
        );

        g.bench_with_input(
            BenchmarkId::from_parameter(format!("{class_name} (No attribute parsing)")),
            &class_bytes,
            |b, bytes| {
                b.iter_batched(
                    || Cursor::new(bytes),
                    |mut cursor| {
                        to_class(&mut cursor, ParsingOption::default())
                            .expect("Parsing fails on benchmarking");
                    },
                    BatchSize::SmallInput,
                );
            },
        );
    }
}

criterion_group!(benches, parsing_classes);
criterion_main!(benches);
