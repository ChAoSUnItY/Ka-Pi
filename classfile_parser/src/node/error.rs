use crate::parse::ParseError;
use thiserror::Error;

pub type NodeResResult<T> = Result<T, NodeResError>;

#[derive(Debug, Error)]
pub enum NodeResError {
    #[error(transparent)]
    ParseFail(#[from] ParseError),
    #[error("failed to parse bytes to string in constant Utf8, given bytes: {0:?}")]
    StringParseFail(Box<[u8]>),
    #[error("unknown constant reference #{0}")]
    UnknownConstantReference(u16),
    #[error("attempt to match {0} in {1:?} but got {2}")]
    MatchOutOfBound(&'static str, Vec<&'static str>, usize),
    #[error("expected referenced constant {0} at #{1} but got {2}")]
    MismatchReferenceConstant(&'static str, u16, &'static str),
    #[error("character {0} cannot be converted into wildcard")]
    InvalidWildcard(char),
    #[error("character {0} cannot be converted into base type")]
    InvalidBaseType(char),
}
