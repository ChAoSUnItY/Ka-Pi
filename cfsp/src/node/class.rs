use serde::{Deserialize, Serialize};

use crate::node::access_flag::ClassAccessFlag;
use crate::node::attribute::AttributeInfo;
use crate::node::constant::ConstantPool;
use crate::node::field::Field;
use crate::node::method::Method;

/// Represents a class file.
///
/// See [4.1 The ClassFile Structure](https://docs.oracle.com/javase/specs/jvms/se20/jvms20.pdf#page=82).
#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub struct Class {
    pub java_version: JavaVersion,
    pub constant_pool_count: u16,
    pub constant_pool: ConstantPool,
    pub access_flag: ClassAccessFlag,
    pub this_class: u16,
    pub super_class: u16,
    pub interfaces_count: u16,
    pub interfaces: Vec<u16>,
    pub fields_count: u16,
    pub fields: Vec<Field>,
    pub methods_count: u16,
    pub methods: Vec<Method>,
    pub attributes_count: u16,
    pub attributes: Vec<AttributeInfo>,
}

impl Class {
    /// Get current class from constant pool.
    pub fn this_class(&self) -> Option<&crate::node::constant::Class> {
        self.constant_pool.get_class(self.this_class)
    }

    /// Get super class from constant pool.
    pub fn super_class(&self) -> Option<&crate::node::constant::Class> {
        self.constant_pool.get_class(self.super_class)
    }

    /// Get interface from constant pool at given index.
    pub fn interface(&self, index: u16) -> Option<&crate::node::constant::Class> {
        self.interfaces
            .get(index as usize)
            .and_then(|interface_index| self.constant_pool.get_class(*interface_index))
    }
}

/// Represents java version documented in specification (combines `major_version` and `minor_version`).
///
/// See [Table 4.1-A](https://docs.oracle.com/javase/specs/jvms/se20/jvms20.pdf#page=83).
#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Hash, Serialize, Deserialize)]
pub enum JavaVersion {
    V1_1,
    V1_2,
    V1_3,
    V1_4,
    V1_5,
    V1_6,
    V1_7,
    V1_8,
    V9,
    V10,
    V11,
    V12,
    V13,
    V14,
    V15,
    V16,
    V17,
    V18,
    V19,
    V20,
    V21,
    Custom { minor: u16, major: u16 },
}

impl JavaVersion {
    #[rustfmt::skip]
    pub const KNOWN_VERSIONS: [JavaVersion; 21] = [
        Self::V1_1, Self::V1_2, Self::V1_3, Self::V1_4, Self::V1_5, Self::V1_6, Self::V1_7, Self::V1_8,
        Self::V9, Self::V10, Self::V11, Self::V12, Self::V13, Self::V14, Self::V15, Self::V16, Self::V17,
        Self::V18, Self::V19, Self::V20, Self::V21,
    ];

    pub fn version(&self) -> u32 {
        match self {
            Self::V1_1 => 3 << 16 | 45,
            Self::V1_2 => 0 << 16 | 46,
            Self::V1_3 => 0 << 16 | 47,
            Self::V1_4 => 0 << 16 | 48,
            Self::V1_5 => 0 << 16 | 49,
            Self::V1_6 => 0 << 16 | 50,
            Self::V1_7 => 0 << 16 | 51,
            Self::V1_8 => 0 << 16 | 52,
            Self::V9 => 0 << 16 | 53,
            Self::V10 => 0 << 16 | 54,
            Self::V11 => 0 << 16 | 55,
            Self::V12 => 0 << 16 | 56,
            Self::V13 => 0 << 16 | 57,
            Self::V14 => 0 << 16 | 58,
            Self::V15 => 0 << 16 | 59,
            Self::V16 => 0 << 16 | 60,
            Self::V17 => 0 << 16 | 61,
            Self::V18 => 0 << 16 | 62,
            Self::V19 => 0 << 16 | 63,
            Self::V20 => 0 << 16 | 64,
            Self::V21 => 0 << 16 | 65,
            JavaVersion::Custom { minor, major } => {
                let minor = minor.to_be_bytes();
                let major = major.to_be_bytes();

                u32::from_be_bytes([minor[0], minor[1], major[0], major[1]])
            }
        }
    }
}

impl From<u32> for JavaVersion {
    fn from(value: u32) -> Self {
        for version in Self::KNOWN_VERSIONS {
            if value == version.version() {
                return version;
            }
        }

        let bits = value.to_be_bytes();

        JavaVersion::Custom {
            minor: u16::from_be_bytes([bits[0], bits[1]]),
            major: u16::from_be_bytes([bits[2], bits[3]]),
        }
    }
}
