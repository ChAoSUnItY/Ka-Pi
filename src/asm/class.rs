use std::cell::RefCell;
use std::rc::Rc;

use crate::asm::byte_vec::{ByteVec, ByteVecImpl};
use crate::asm::field::{FieldVisitor, FieldWriter};
use crate::asm::method::{MethodVisitor, MethodWriter};
use crate::asm::opcodes::{
    AccessFlag, ClassAccessFlag, FieldAccessFlag, JavaVersion, MethodAccessFlag,
};
use crate::asm::symbol::{Constant, SymbolTable};
use crate::error::KapiResult;

pub trait ClassVisitor {
    type MethodVisitor: MethodVisitor + Sized;
    type FieldVisitor: FieldVisitor + Sized;

    fn visit_method<F>(
        &mut self,
        access_flags: F,
        name: &str,
        descriptor: &str,
    ) -> KapiResult<Self::MethodVisitor>
    where
        F: IntoIterator<Item = MethodAccessFlag>;

    fn visit_field<F>(
        &mut self,
        access_flags: F,
        name: &str,
        descriptor: &str,
    ) -> KapiResult<Self::FieldVisitor>
    where
        F: IntoIterator<Item = FieldAccessFlag>;

    fn visit_end(&self) {}
}

pub struct ClassWriter {
    byte_vec: Rc<RefCell<ByteVecImpl>>,
    symbol_table: Rc<RefCell<SymbolTable>>,
    version: JavaVersion,
    access_flags: Vec<ClassAccessFlag>,
    this_class_index: u16,
    super_class_index: u16,
    interface_indices: Vec<u16>,
    fields: Vec<Rc<RefCell<ByteVecImpl>>>,
    methods: Vec<Rc<RefCell<ByteVecImpl>>>,
}

impl ClassWriter {
    pub fn new_class_writer<F, I>(
        version: JavaVersion,
        access_flags: F,
        class_name: &str,
        super_class: &str,
        interfaces: I,
    ) -> Self
    where
        F: IntoIterator<Item = ClassAccessFlag>,
        I: IntoIterator<Item = String>,
    {
        let mut symbol_table = SymbolTable::default();

        let this_class_index = symbol_table.add_class(class_name);
        let super_class_index = symbol_table.add_class(super_class);
        let interface_indices = interfaces
            .into_iter()
            .map(|interface| symbol_table.add_class(&interface))
            .collect::<Vec<_>>();

        Self {
            byte_vec: Rc::new(RefCell::new(ByteVecImpl::new())),
            symbol_table: Rc::new(RefCell::new(symbol_table)),
            version,
            access_flags: access_flags.into_iter().collect(),
            this_class_index,
            super_class_index,
            interface_indices,
            fields: Vec::new(),
            methods: Vec::new(),
        }
    }

    pub fn bytecode(&self) -> Vec<u8> {
        self.byte_vec.borrow().clone()
    }
}

impl ClassVisitor for ClassWriter {
    type MethodVisitor = MethodWriter;
    type FieldVisitor = FieldWriter;

    fn visit_method<F>(
        &mut self,
        access_flags: F,
        name: &str,
        descriptor: &str,
    ) -> KapiResult<Self::MethodVisitor>
    where
        F: IntoIterator<Item = MethodAccessFlag>,
    {
        let method_byte_vec = Rc::new(RefCell::new(ByteVecImpl::with_capacity(8)));

        self.methods.push(method_byte_vec.clone());

        MethodWriter::new(
            &method_byte_vec,
            &self.symbol_table,
            access_flags,
            name,
            descriptor,
        )
    }

    fn visit_field<F>(
        &mut self,
        access_flags: F,
        name: &str,
        descriptor: &str,
    ) -> KapiResult<Self::FieldVisitor>
    where
        F: IntoIterator<Item = FieldAccessFlag>,
    {
        let field_byte_vec = Rc::new(RefCell::new(ByteVecImpl::with_capacity(8)));

        self.fields.push(field_byte_vec.clone());

        FieldWriter::new(
            &field_byte_vec,
            &self.symbol_table,
            access_flags,
            name,
            descriptor,
        )
    }

    fn visit_end(&self)
    where
        Self: Sized,
    {
        let Self {
            byte_vec,
            symbol_table,
            version,
            access_flags,
            this_class_index,
            super_class_index,
            interface_indices,
            fields,
            methods,
        } = self;

        let mut byte_vec = byte_vec.borrow_mut();
        let mut symbol_table = symbol_table.borrow_mut();

        byte_vec.put_u8s(&[0xCA, 0xFE, 0xBA, 0xBE]); // magic number
        byte_vec.put_u8s(&(*version as u32).to_be_bytes()); // major version, minor version

        byte_vec.put_be(symbol_table.constants.len() as u16 + 1); // constant pool length
        for constant in &symbol_table.constants {
            byte_vec.put_be(constant.tag() as u8);

            match constant {
                Constant::Class { name_index } => {
                    byte_vec.put_be(*name_index);
                }
                Constant::FieldRef {
                    class_index,
                    name_and_type_index,
                } => {
                    byte_vec.put_be(*class_index);
                    byte_vec.put_be(*name_and_type_index);
                }
                Constant::MethodRef {
                    class_index,
                    name_and_type_index,
                } => {
                    byte_vec.put_be(*class_index);
                    byte_vec.put_be(*name_and_type_index);
                }
                Constant::InterfaceMethodRef {
                    class_index,
                    name_and_type_index,
                } => {
                    byte_vec.put_be(*class_index);
                    byte_vec.put_be(*name_and_type_index);
                }
                Constant::String { string_index } => byte_vec.put_be(*string_index),
                Constant::Integer { bytes } => {
                    byte_vec.extend_from_slice(bytes);
                }
                Constant::Float { bytes } => {
                    byte_vec.extend_from_slice(bytes);
                }
                Constant::Long {
                    high_bytes,
                    low_bytes,
                } => {
                    byte_vec.extend_from_slice(high_bytes);
                    byte_vec.extend_from_slice(low_bytes);
                }
                Constant::Double {
                    high_bytes,
                    low_bytes,
                } => {
                    byte_vec.extend_from_slice(high_bytes);
                    byte_vec.extend_from_slice(low_bytes);
                }
                Constant::NameAndType {
                    name_index,
                    type_index,
                } => {
                    byte_vec.put_be(*name_index);
                    byte_vec.put_be(*type_index);
                }
                Constant::Utf8 { data } => {
                    byte_vec.put_utf8(&data).unwrap();
                }
                Constant::MethodHandle {
                    reference_kind,
                    reference_index,
                } => {
                    byte_vec.put_be(*reference_kind);
                    byte_vec.put_be(*reference_index);
                }
                Constant::MethodType { descriptor } => {
                    byte_vec.put_be(*descriptor);
                }
                Constant::Dynamic {
                    bootstrap_method_attr_index,
                    name_and_type_index,
                } => {
                    byte_vec.put_be(*bootstrap_method_attr_index);
                    byte_vec.put_be(*name_and_type_index);
                }
                Constant::InvokeDynamic {
                    bootstrap_method_attr_index,
                    name_and_type_index,
                } => {
                    byte_vec.put_be(*bootstrap_method_attr_index);
                    byte_vec.put_be(*name_and_type_index);
                }
                Constant::Module { name_index } => {
                    byte_vec.put_be(*name_index);
                }
                Constant::Package { name_index } => {
                    byte_vec.put_be(*name_index);
                }
            }
        }

        byte_vec.put_be(access_flags.fold_flags()); // access flags
        byte_vec.put_be(*this_class_index); // this class
        byte_vec.put_be(*super_class_index); // super class
        byte_vec.put_be(interface_indices.len() as u16); // interfaces length

        for interface_index in interface_indices {
            byte_vec.put_be(*interface_index);
        }

        byte_vec.put_be(fields.len() as u16); // fields length
        for field_segment in fields {
            byte_vec.put_u8s(&field_segment.borrow()[..]);
        }

        byte_vec.put_be(methods.len() as u16); // methods length
        for method_segment in methods {
            byte_vec.put_u8s(&method_segment.borrow()[..]);
        }

        byte_vec.put_be(symbol_table.attributes.len() as u16); // attributes length
                                                               // TODO: implement attributes
    }
}
