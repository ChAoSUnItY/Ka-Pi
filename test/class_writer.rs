use std::fs::File;
use std::io::Write;
use insta::assert_yaml_snapshot;
use ka_pi::asm::generate::class::ClassWriter;
use ka_pi::asm::node::access_flag::{ClassAccessFlag, FieldAccessFlag, MethodAccessFlag};
use ka_pi::asm::node::class::JavaVersion;
use ka_pi::asm::parse::to_class;
use ka_pi::error::KapiResult;

#[test]
fn validate_class_writer_output() -> KapiResult<()> {
    let mut class_writer = ClassWriter::new(
        JavaVersion::V17,
        vec![ClassAccessFlag::Super, ClassAccessFlag::Public],
        "Main",
        "java/lang/Object",
        vec![],
    );

    class_writer.write_field(
        vec![FieldAccessFlag::Public, FieldAccessFlag::Static],
        "field",
        "Z",
        |field| Ok(field),
    )?;
    class_writer.write_method(
        vec![MethodAccessFlag::Public, MethodAccessFlag::Static],
        "method",
        "()Z",
        |method| Ok(method),
    )?;

    let bytes = class_writer.write_output()?;

    assert!(!bytes.is_empty());

    let class = to_class(&bytes)?;

    assert_yaml_snapshot!(class);

    Ok(())
}
