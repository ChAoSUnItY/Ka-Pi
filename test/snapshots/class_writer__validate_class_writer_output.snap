---
source: test/class_writer.rs
expression: class
---
java_version: V17
constant_pool_count: 9
constant_pool:
  len: 9
  entries:
    1:
      Utf8:
        length: 5
        bytes:
          - 102
          - 105
          - 101
          - 108
          - 100
    2:
      Utf8:
        length: 1
        bytes:
          - 90
    3:
      Utf8:
        length: 6
        bytes:
          - 109
          - 101
          - 116
          - 104
          - 111
          - 100
    4:
      Utf8:
        length: 3
        bytes:
          - 40
          - 41
          - 90
    5:
      Utf8:
        length: 4
        bytes:
          - 77
          - 97
          - 105
          - 110
    6:
      Class:
        name_index: 5
    7:
      Utf8:
        length: 16
        bytes:
          - 106
          - 97
          - 118
          - 97
          - 47
          - 108
          - 97
          - 110
          - 103
          - 47
          - 79
          - 98
          - 106
          - 101
          - 99
          - 116
    8:
      Class:
        name_index: 7
access_flags:
  - Public
  - Super
this_class: 6
super_class: 8
interfaces_count: 0
interfaces: []
fields_count: 1
fields:
  - access_flags:
      - Public
      - Static
    name_index: 1
    descriptor_index: 2
    attribute_infos_len: 0
    attribute_infos: []
methods_count: 1
methods:
  - access_flags:
      - Public
      - Static
    name_index: 3
    descriptor_index: 4
    attribute_infos_len: 0
    attribute_infos: []
attributes_count: 0
attributes: []

